File: Week6/sentimental-credit/credit.py
# TODO
from cs50 import get_string

number = get_string("Number: ")


def main():
    if not luhns_algorithm([int(i) for i in number]):
        print("INVALID")

    provider = int(number[:2])
    numlen = len(number)
    if (provider == 34 or provider == 37) and numlen == 15:
        print("AMEX")
    elif 55 >= provider >= 51 and numlen == 16:
        print("MASTERCARD")
    elif 49 >= provider >= 40 and (numlen == 13 or numlen == 16):
        print("VISA")
    else:
        print("INVALID")


def luhns_algorithm(card_number: list[int]) -> bool:
    check_sum = 0

    for i, number in enumerate(card_number[::-1], 1):
        number = int(number)

        if i % 2:
            number *= 2
            check_sum += number % 10 + number // 10
        else:
            check_sum += number

    return check_sum % 10 == 0


if __name__ == "__main__":
    main()

File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool luhnAlgorithm(long cardNumber);
string cardChecker(long cardNumber);

int main(void)
{
    long cardNumber;
    int length;

    // Asks for card number
    cardNumber = get_long("Enter a card number: ");

    printf("%s", cardChecker(cardNumber));
}

// Checks if card is valid
string cardChecker(long cardNumber)
{
    if (!luhnAlgorithm(cardNumber))
    {
        return "INVALID\n";
    }

    // Gets card length and provider
    int cardLength = (int)floor(1 + log10(cardNumber));
    int provider = cardNumber / pow(10, cardLength - 2);

    // Returns right provider if card is valid
    if (provider == 34 || provider == 37)
    {
        if (cardLength == 15)
        {
            return "AMEX\n";
        }
    }
    else if (provider >= 51 && provider <= 55)
    {
        if (cardLength == 16)
        {
            return "MASTERCARD\n";
        }
    }
    else if (provider >= 40 && provider  <= 49)
    {
        if (cardLength == 13 || cardLength == 16)
        {
            return "VISA\n";
        }
    }

    return "INVALID\n";
}

// Own implementation of Luhn Algorithm
bool luhnAlgorithm(long cardNumber)
{
    // Cal
    int cardLength = (int)floor(1 + log10(cardNumber));

    int sum = 0;

    bool isSecond = false;
    int number;
    // Calculation of sum
    for (int i = 0; i < cardLength; i++)
    {
        number = cardNumber % 10;
        if (isSecond)
        {
            number *= 2;
            if (number < 10)
            {
                sum += number;
            }
            else
            {
                sum += number % 10;
                sum += number / 10;
            }
        }
        else
        {
            sum += number;
        }

        cardNumber /= 10;
        isSecond = !isSecond;
    }

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}